Automattic/woocommerce-payments
# Enable payment methods when appropriate

Fixes #2192 

#### Changes proposed in this Pull Request


### Refactoring UPE payment methods by extracting them from the gateway

> The goal of this issue is to provide a means to enable/disable certain pay methods within the UPE depending on the store and cart. Since Giropay is the first pay method outside of credit card in UPE we'll be using, this will be laying the foundation for the other pay methods as they become available. While the leading thought is to create a separate class (NOT a gateway) for these pay methods to centralize the functionality, other approaches can be explored as well.

This PR separates the internal logic for any UPE payment methods from the gateway itself. This makes the gateway ignorant of any payment method logic, removing any branching logic or switch cases dependent on the payment method type (Stripe key). This also allows us to define logic about whether payment methods can be saved or shown when subscription products are present in one place on a per payment method basis. 

Since all payment methods also inherit most functionality from a base class, you'll see that this will mean that we will be able to add new UPE payment methods very easily, simply by creating a lightweight new class and making the gateway aware of its existence. 







#### Testing instructions



1. Please pull these changes, as well as the latest version of the Payments Server and the WCPay Dev Utils plugin. Using the latter plugin, please ensure that `Enabled grouped settings`, `Enable UPE checkout`, `Enable UPE settings changes`, and `Add UPE additional payment methods` are all checked.
2. Either add an item from your store to cart and proceed to the regular checkout or create an order and proceed to a customer's Order Pay page to complete payment for the order (or try both, if you're feeling dangerous). Note that if you try to use the customer's Order Pay page to pay for an Order containing a subscription item (create a new Subscription, then create a parent order from this subscription, and finally visit the Order Pay page for this order), you will actually be sent to the regular checkout page instead of the Order Pay page. Anyways, from either of these pages, ensure that the regular payment flow still works and then you can focus on the following new features.
3. Within the UPE element, if you toggle between payment methods, check that the `Save payment information` checkbox (when logged into an account) is only visible and selectable when the payment method is reusable (e.g. Card and Sofort).
4. If the cart at checkout contains a Subscription product, check that only reusable payment methods are present within the UPE.



-------------------
